src/ray/raylet/scheduling/cluster_resource_scheduler.cc

```diff
@@ -402,36 +402,41 @@ const NodeResources &ClusterResourceScheduler::GetLocalNodeResources() const {
   return node_it->second.GetLocalView();
 }
 
-int64_t ClusterResourceScheduler::NumNodes() { return nodes_.size(); }
+int64_t ClusterResourceScheduler::NumNodes() const { return nodes_.size(); }
 
-void ClusterResourceScheduler::AddLocalResource(const std::string &resource_name,
-                                                double resource_total) {
-  string_to_int_map_.Insert(resource_name);
-  int64_t resource_id = string_to_int_map_.Get(resource_name);
+const StringIdMap ClusterResourceScheduler::GetStringIdMap() const {
+  return string_to_int_map_;
+}
 
-  if (local_resources_.custom_resources.contains(resource_id)) {
-    FixedPoint total(resource_total);
-    auto &instances = local_resources_.custom_resources[resource_id];
-    instances.total[0] += total;
-    instances.available[0] += total;
-    auto local_node_it = nodes_.find(local_node_id_);
-    RAY_CHECK(local_node_it != nodes_.end());
-    auto &capacity =
-        local_node_it->second.GetMutableLocalView()->custom_resources[resource_id];
-    capacity.available += total;
-    capacity.total += total;
+void ClusterResourceScheduler::AddLocalResourceInstances(
+    const std::string &resource_name, const std::vector<FixedPoint> &instances) {
+  ResourceInstanceCapacities *node_instances;
+  local_resources_.predefined_resources.resize(PredefinedResources_MAX);
+  if (kCPU_ResourceLabel == resource_name) {
+    node_instances = &local_resources_.predefined_resources[CPU];
```

Comment:
Why do you use `&local_resources_.` here? (Why do you use & in the beginning of this attribute?)

Reply:
To get a pointer. In an ideal world, c++ would let us do `ResourceInstanceCapacities &node_instances` then initialize it in the if statements, but this is the next best thing.